crypto/keyring/keyring.go

```diff
@@ -842,26 +837,166 @@ func (ks keystore) existsInDb(info Info) (bool, error) {
-	info, err := NewMultiInfo(name, pub)
+// writeMultisigKey investigate where thisf function is called maybe remove it
+func (ks keystore) writeMultisigKey(name string, pk types.PubKey) (*Record, error) {
+	k, err := NewMultiRecord(name, pk)
 	if err != nil {
 		return nil, err
 	}
-	if err = ks.writeInfo(info); err != nil {
+
+	return k, ks.writeRecord(k)
+}
+
+func (ks keystore) MigrateAll() (bool, error) {
+	var migrated bool
+	keys, err := ks.db.Keys()
+	if err != nil {
+		return migrated, err
+	}
+
+	if len(keys) == 0 {
+		fmt.Println("no keys available for migration")
+		return migrated, nil
+	}
+
+	for _, key := range keys {
+		if strings.Contains(key, addressSuffix) {
+			continue
+		}
+
+		migrated, err = ks.migrate(key)
+		if err != nil {
+			fmt.Printf("migrate err: %q", err)
+			continue
+		}
+	}
+
+	return migrated, nil
+}
+
+// for one key
+func (ks keystore) migrate(key string) (bool, error) {
+	var migrated bool
+	item, err := ks.db.Get(key)
+	if err != nil {
+		return migrated, wrapKeyNotFound(err, key)
+	}
+
+	if len(item.Data) == 0 {
+		return migrated, sdkerrors.Wrap(sdkerrors.ErrKeyNotFound, key)
+	}
+
+	// 2.try to deserialize using proto, if good then continue, otherwise try to deserialize using amino
+	if _, err := ks.protoUnmarshalRecord(item.Data); err == nil {
+		return migrated, nil
+	}
+
+	LegacyInfo, err := unMarshalLegacyInfo(item.Data)
+	if err != nil {
+		return migrated, fmt.Errorf("unable to unmarshal item.Data, err: %w", err)
+	}
+
+	// 4.serialize info using proto
+	k, err := ks.convertFromLegacyInfo(LegacyInfo)
+	if err != nil {
+		return migrated, fmt.Errorf("convertFromLegacyInfo, err - %s", err)
+	}
+
+	serializedRecord, err := ks.protoMarshalRecord(k)
+	if err != nil {
+		return migrated, fmt.Errorf("unable to serialize record, err - %w", err)
+	}
+
+	item = keyring.Item{
+		Key:         key,
+		Data:        serializedRecord,
+		Description: "SDK kerying version",
+	}
+	// 5.overwrite the keyring entry with
+	if err := ks.SetItem(item); err != nil {
+		return migrated, fmt.Errorf("unable to set keyring.Item, err: %w", err)
+	}
+
+	return !migrated, nil
+}
+
+func (ks keystore) protoUnmarshalRecord(bz []byte) (*Record, error) {
+	k := new(Record)
+	if err := ks.cdc.Unmarshal(bz, k); err != nil {
 		return nil, err
 	}
 
-	return info, nil
+	return k, nil
+}
+
+func (ks keystore) protoMarshalRecord(k *Record) ([]byte, error) {
+	return ks.cdc.Marshal(k)
+}
+
+func (ks keystore) SetItem(item keyring.Item) error {
+	return ks.db.Set(item)
+}
+
+func (ks keystore) convertFromLegacyInfo(info LegacyInfo) (*Record, error) {
+	if info == nil {
+		return nil, errors.New("unable to convert LegacyInfo to Record cause info is nil")
+	}
+
+	name := info.GetName()
+	pk := info.GetPubKey()
+
+	switch info.GetType() {
+	case TypeLocal:
+		priv, err := exportPrivateKeyFromLegacyInfo(info)
+		if err != nil {
+			return nil, err
+		}
+
+		k, err := NewLocalRecord(name, priv, pk)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
how about returning directly:
```
return NewLocalRecord(name, priv, pk)
```
Same for the cases below.

Reply:
Done!